packages/workspace/src/workspace/nacl_files/nacl_files_source.ts

```diff
@@ -112,36 +111,47 @@ const getTypeOrContainerTypeID = (typeElem: TypeElement): ElemID => (isContainer
   ? getDeepInnerType(typeElem).elemID
   : typeElem.elemID)
 
-const getElementReferenced = (element: Element): ElemID[] => {
-  const referenced: ElemID[] = []
+const getElementReferenced = (element: Element): Set<string> => {
+  const referenced = new Set<string>()
   const transformFunc = ({ value, field, path }: TransformFuncArgs): Value => {
     if (field && path && !isIndexPathPart(path.name)) {
-      referenced.push(getTypeOrContainerTypeID(field.type))
+      referenced.add(getTypeOrContainerTypeID(field.type).getFullName())
     }
     if (isReferenceExpression(value)) {
       const { parent, path: valueIDPath } = value.elemId.createTopLevelParentID()
       const nestedIds = valueIDPath.map((_p, index) => parent.createNestedID(
-        value.elemId.idType,
+        ...(value.elemId.idType !== parent.idType ? [value.elemId.idType] : []),
         ...valueIDPath.slice(0, index + 1)
       ))
-      referenced.push(parent, ...nestedIds)
+      referenced.add(parent.getFullName())
+      nestedIds.forEach(id => referenced.add(id.getFullName()))
     }
     return value
   }
 
   if (isObjectType(element)) {
-    referenced.push(...Object.values(element.fields)
-      .map(field => getTypeOrContainerTypeID(field.type)))
+    Object.values(element.fields)
+      .map(field => getTypeOrContainerTypeID(field.type))
+      .forEach(id => referenced.add(id.getFullName()))
   }
   if (isInstanceElement(element)) {
-    referenced.push(element.type.elemID)
+    referenced.add(element.type.elemID.getFullName())
   }
-  referenced.push(...Object.values(element.annotationTypes)
-    .map(anno => getTypeOrContainerTypeID(anno)))
+  Object.values(element.annotationTypes)
+    .map(anno => getTypeOrContainerTypeID(anno))
+    .forEach(id => referenced.add(id.getFullName()))
   if (!isContainerType(element) && !isVariable(element)) {
     transformElement({ element, transformFunc, strict: false })
   }
-  return _.uniq(referenced)
+  return referenced
+}
+
+const getElementsReferences = (elements: Element[]): ElemID[] => {
+  const referenced = new Set<string>()
+  elements.forEach(elem => {
+    getElementReferenced(elem).forEach(r => referenced.add(r))
+  })
```

Comment:
style nit - avoid side effects when not needed
```suggestion
  const referenced = new Set(wu(elements).map(getElementReferenced))
```